src/master-playlist-controller.js

```diff
@@ -0,0 +1,519 @@
+      mediaSource: this.mediaSource,
+      currentTime: this.currentTimeFunc,
+      withCredentials: this.withCredentials
+    });
+
+    this.hlsHandler = hlsHandler;
+    this.hlsHandler.selectPlaylist = this.hlsHandler.selectPlaylist || selectPlaylist;
+
+    if (!url) {
+      throw new Error('A non-empty playlist URL is required');
+    }
+
+    this.masterPlaylistLoader_ = new PlaylistLoader(url, this.withCredentials);
+    this.hlsHandler.playlists = this.masterPlaylistLoader_;
+
+    this.masterPlaylistLoader_.on('loadedmetadata', () => {
+      // if this isn't a live video and preload permits, start
+      // downloading segments
+      if (this.masterPlaylistLoader_.media().endList &&
+          this.hlsHandler.tech_.preload() !== 'metadata' &&
+          this.hlsHandler.tech_.preload() !== 'none') {
+        this.mainSegmentLoader_.playlist(this.masterPlaylistLoader_.media());
+        this.mainSegmentLoader_.load();
+      }
+
+      this.setupFirstPlay();
+      this.hlsHandler.tech_.trigger('loadedmetadata');
+    });
+
+    this.masterPlaylistLoader_.on('loadedplaylist', () => {
+      let updatedPlaylist = this.masterPlaylistLoader_.media();
+      let seekable;
+
+      if (!updatedPlaylist) {
+        // select the initial variant
+        this.masterPlaylistLoader_.media(this.hlsHandler.selectPlaylist());
+        return;
+      }
+
+      this.mainSegmentLoader_.playlist(updatedPlaylist);
+      this.updateDuration(this.masterPlaylistLoader_.media());
+
+      // update seekable
+      seekable = this.hlsHandler.seekable();
+      if (this.hlsHandler.duration() === Infinity && seekable.length !== 0) {
+        this.mediaSource.addSeekableRange_(seekable.start(0), seekable.end(0));
+      }
+    });
+
+    this.masterPlaylistLoader_.on('error', () => {
+      this.blacklistCurrentPlaylist_(this.masterPlaylistLoader_.error);
+    });
+
+    this.masterPlaylistLoader_.on('mediachanging', () => {
+      this.mainSegmentLoader_.pause();
+    });
+
+    this.masterPlaylistLoader_.on('mediachange', () => {
+      this.mainSegmentLoader_.abort();
+      this.mainSegmentLoader_.load();
+      this.hlsHandler.tech_.trigger({
+        type: 'mediachange',
+        bubbles: true
+      });
+    });
+
+    this.mainSegmentLoader_.on('progress', () => {
+      // figure out what stream the next segment should be downloaded from
+      // with the updated bandwidth information
+      this.hlsHandler.bandwidth = this.mainSegmentLoader_.bandwidth;
+      this.masterPlaylistLoader_.media(this.hlsHandler.selectPlaylist());
+
+      this.hlsHandler.tech_.trigger('progress');
+    });
+
+    this.mainSegmentLoader_.on('error', () => {
+      this.blacklistCurrentPlaylist_(this.mainSegmentLoader_.error());
+    });
+  }
+
+  load() {
+    this.mainSegmentLoader_.load();
+    if (this.audioPlaylistLoader_) {
+      this.audioSegmentLoader_.load();
+    }
+  }
+
+  loadAlternateAudioPlaylist(playlistLoader) {
```

Comment:
I see what you're saying now. It lets us encapsulate the selection logic (rather, the listening to selected audio tracks) within HLS, instead of MPC having to keep track of that as an external reference to the player's audio track list change events.

That makes a lot of sense, and I'll modify it to allow for that interaction.